server/client.go

```diff
@@ -1966,6 +2001,62 @@ func (c *client) canSubscribe(subject string) bool {
 	return allowed
 }
 
+func queueMatches(queue string, qsubs [][]*subscription) bool {
+	if len(qsubs) == 0 {
+		return true
+	}
+	for _, qsub := range qsubs {
+		qs := qsub[0]
+		qname := string(qs.queue)
+
+		// NOTE: '*' and '>' tokens can also be valid
+		// queue names so we first check against the
+		// literal name.  e.g. v1.* == v1.*
+		if queue == qname || (subjectHasWildcard(qname) && subjectIsSubsetMatch(queue, qname)) {
+			return true
+		}
+	}
+	return false
+}
+
+func (c *client) canQueueSubscribe(subject, queue string) bool {
+	if c.perms == nil {
+		return true
+	}
+
+	allowed := true
+	if c.perms.sub.allow != nil {
+		r := c.perms.sub.allow.Match(subject)
+
+		// If there is an explicit sub permission it allows any
+		// queue subscription unless there are explicit rules
+		// involving queue subscriptions.
+		subAllowed := len(r.psubs) != 0
+		allQueuesAllowed := len(r.qsubs) == 0
+
+		if allowed = subAllowed && allQueuesAllowed; !allowed {
+			// If the queue matches, then the queue name is on the allow list.
+			// Therefore, we ARE allowed.
+			allowed = queueMatches(queue, r.qsubs)
+		}
+	}
+
+	// Check whether there is an explicit deny against this queue subscription.
+	if allowed && c.perms.sub.deny != nil {
+		r := c.perms.sub.deny.Match(subject)
+		subAllowed := len(r.psubs) == 0
```

Comment:
If I was not wrong for the "allow" case, I think we could apply the same idea here.